### PR TITLE
complete EdgeDeployment targeting

### PR DESCRIPTION
**Pull request by ydayagi**
https://issues.redhat.com/browse/ECOPROJECT-308
Signed-off-by: Yaron Dayagi <ydayagi@redhat.com>

### CODE REVIEW

File: controllers/edgedeployment_controller.go [hunk 1]
```diff
@@ -367,3 +381,60 @@ func splitEdgeDevices(edgeDevices []managementv1alpha1.EdgeDevice, splitSize uin
...
+
+	err := r.EdgeDeploymentRepository.Patch(ctx, edgeDeployment, edgeDeploymentCopy)
+	return err
+}
+
+func isSelectorLabel(label string) bool {
+	return strings.HasPrefix(label, selectorLabelPrefix)
+}
+
+func CreateSelectorLabel(label string) string {
```
Comment (jakub-dzon) [inline controllers/edgedeployment_controller.go:402]: This should be private.

File: controllers/edgedeployment_controller.go [hunk 2]
```diff
@@ -367,3 +381,60 @@ func splitEdgeDevices(edgeDevices []managementv1alpha1.EdgeDevice, splitSize uin
...
+	return selectorLabelPrefix + label
+}
+
+func UpdateSelectorLabels(edgeDeployments ...*managementv1alpha1.EdgeDeployment) {
```
Comment (jakub-dzon) [inline controllers/edgedeployment_controller.go:406]: This should be private.

**Review by jakub-dzon — COMMENTED** [review]
Good work! 
I just have a couple of cosmetic comments.

File: controllers/edgedeployment_controller.go (hunk 2 above)
Comment (ydayagi) [reply controllers/edgedeployment_controller.go:406]: I use it in tests. Tests are a separate package.

File: controllers/edgedeployment_controller.go (hunk 1 above)
Comment (ydayagi) [reply controllers/edgedeployment_controller.go:402]: I use it in tests. Tests are a separate package.

File: controllers/edgedeployment_controller.go (hunk 2 above)
Comment (ydayagi) [reply controllers/edgedeployment_controller.go:406]: I use it in tests. Tests are a separate package.

File: controllers/edgedeployment_controller.go (hunk 1 above)
Comment (jakub-dzon) [reply controllers/edgedeployment_controller.go:402]: In that case it looks like a good candidate for moving to https://github.com/jakub-dzon/k4e-operator/blob/8081cb355b8c1daa5af3e15cb85ea92ea00e2093/internal/labels/labels.go

File: controllers/edgedeployment_controller.go (hunk 2 above)
Comment (jakub-dzon) [reply controllers/edgedeployment_controller.go:406]: OK, that makes sense in this case.

File: controllers/edgedeployment_controller.go [hunk 3]
```diff
@@ -367,3 +381,60 @@ func splitEdgeDevices(edgeDevices []managementv1alpha1.EdgeDevice, splitSize uin
...
+
+	if (len(newLabels) == 0 && edgeDeployment.Labels == nil) || reflect.DeepEqual(newLabels, edgeDeployment.Labels) {
+		return nil
+	}
+
+	err := r.EdgeDeploymentRepository.Patch(ctx, edgeDeployment, edgeDeploymentCopy)
+	return err
+}
+
+func isSelectorLabel(label string) bool {
```
Comment (jakub-dzon) [inline controllers/edgedeployment_controller.go:398]: This one also could be moved to https://github.com/jakub-dzon/k4e-operator/blob/8081cb355b8c1daa5af3e15cb85ea92ea00e2093/internal/labels/labels.go

File: controllers/edgedeployment_controller.go (hunk 1 above)
Comment (ydayagi) [reply controllers/edgedeployment_controller.go:402]: Sure, I'll move it. But, I think we have too many files and packages. I worry that this way people won't find the code they can reuse. I personally, prefer larger files and packages that include a lot. Unless there's some logic for the separation.

File: controllers/edgedeployment_controller.go (hunk 3 above)
Comment (ydayagi) [reply controllers/edgedeployment_controller.go:398]: ok

**Review by jakub-dzon — APPROVED** [review]

File: controllers/edgedevicelabels_controller.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+//+kubebuilder:rbac:groups=management.k4e.io,resources=edgedevices,verbs=get;watch;patch
+//+kubebuilder:rbac:groups=management.k4e.io,resources=edgedevices/status,verbs=get;patch
+//+kubebuilder:rbac:groups=management.k4e.io,resources=edgedeployments,verbs=list
+
+func (r *EdgeDeviceLabelsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName("labels")
+
+	logger.Info("Reconciling")
```
[the remark this answers is not in the document]
Comment (ydayagi) [reply controllers/edgedevicelabels_controller.go:49]: if you look at the controller log you'll see that it prints all these details. It prints the resource type/kind, request fields (name and namespace), name of controller/reconciler (in our case 'labels').
you end up with a log line that has it all.
`{"level":"info","ts":1641400210.3005347,"logger":"controller-runtime.manager.controller.edgedevice.labels","msg":"Reconciling","reconciler group":"management.k4e.io","reconciler kind":"EdgeDevice","name":"b9fb8329-6623-4e13-800e-c3839998f8c5","namespace":"default"}`

File: controllers/edgedeployment_controller.go [hunk 5]
```diff
@@ -87,6 +88,13 @@ func (r *EdgeDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{Requeue: true}, nil
 	}
 
+	if edgeDeployment.DeletionTimestamp == nil {
+		err = r.updateLabelsFromSelector(ctx, edgeDeployment)
```
[the remark this answers is not in the document]
Comment (ydayagi) [reply controllers/edgedeployment_controller.go:92]: as we discussed, I'll test and fix it if our theory is correct. Thanks again for this finding.

File: controllers/edgedevicelabels_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ydayagi) [reply controllers/edgedevicelabels_controller.go:49]: fixed

File: controllers/edgedeployment_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ydayagi) [reply controllers/edgedeployment_controller.go]: fixed

**Review by masayag — APPROVED** [review]

**Review by pkliczewski — APPROVED** [review]

**Pull request merged**